### PR TITLE
[fud] Cleanup the executor interface

### PR DESCRIPTION
**Pull request by rachitnigam**
The code that executes stages and steps is intertwined with the formatting and the profiling code. This attempts to clean that up by defining new `Executor` objects which make use of `__enter__` and `__exit__` magic methods in python to expose a cleaner interface.

The broader goal of this work is to make the executor a reusable component which will allow us to use `fud` as a library when building other tools.

### CODE REVIEW

File: fud/fud/executor.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+        assert (
+            self._current_time is not None
+        ), "Attempt to end measurement before it starts"
+        t = self._current_time
+        self._current_time = None
+        return time.time() - t
+
+
+class DummyProfiler:
+    """
```
Comment (rachitnigam) [inline fud/fud/executor.py:48]: @sampsyo is there a more pythonic way to define shims like these? I don't the code for the Executor to have `if self.profiler is not None` everywhere so I did this.

**Comment by rachitnigam** [comment]
The next step to build on top of this PR is removing the distinction between stages and steps in the executor so that it only executes functions given to it. This will allow steps to call other *stages*. This will be helpful when a step needs another file to compiled using `fud` itself.

For example, in the symbolic evaluator, we may get a specification file written in Dahlia that needs to be verified against a systolic array. In this case, both files need to be compiled to racket before they can be verified. The aforementioned change would allow us to call fud from within the step.

File: fud/fud/executor.py (hunk 1 above)
Comment (sampsyo) [reply fud/fud/executor.py:48]: I think this is a pretty good route! No shame in this.

I think another good alternative would be to have a bool flag in the `Profiler` class that turns it into a no-op. But this dummy-substitute way seems perfectly reasonable to me!

**Comment by sampsyo** [comment]
Wow, cool. Steps invoking stages would be quite a thing.

**Comment by rachitnigam** [comment]
Hopefully this doesn't break anything

**Pull request merged**